### ISSUE TITLE
remapping cursor movement keys in insert mode

### ISSUE THREAD

**Issue by muellerto** [VSCodeVim/Vim#3863]
I'm generally not a friend of cursor movement in insert mode. In the very most cases I want to type a command (cw, ct..., cf..., yw, Y, p, P, ...) after moving the cursor, this is my experience for 30 years now. But many vi emulations stay in insert mode after pressing a cursor movement key. I find this more obstructive than helpful.

So I normally remap the cursor keys in insert mode as follows:
`<left> --> <esc>`
`<right> --> <esc> l`
`<up> --> <esc> l k`
`<down> --> <esc> l j`
(also some more keys like Home, End, PgUp and PgDn)

In VSCodeVim/Vim this should look like this:
`"vim.insertModeKeyBindingsNonRecursive": [`
`{"before": ["<left>"], "after": ["<esc>"]},`
`{"before": ["<right>"], "after": ["<esc>", "l"]},`
`{"before": ["<up>"], "after": ["<esc>", "l", "k"]},`
`{"before": ["<down>"], "after": ["<esc>", "l", "j"]}`
`],`

But it doesn't work very well. It's just not reliable, sometimes the insert mode is terminated, in the most cases it is not. Sometimes the cursor moves to the expected position, sometimes it moves somewhere else. Why? I also see a visible selection being displayed for a short moment. Why???

What I expect is: `<esc>` should always and under all circumstances terminate the insert mode and switch to the normal mode, where keys like `h` or `j` or `k` or `l` should do a reliable cursor movement.

I also detected that cursor movement in normal mode is also not really good, especially up and down. Try this in your `settings.json` file: go onto the first letter of any line, this is the character right of a " and then move up and down using the arrow keys or `j` and `k`. The cursor should always stay in the same column but it doesn't. On some lines the cursor jumps three characters to the right. Why? These lines seem to be always the same lines, so they must have a hidden criterion which leads to the jump, but what is it?

My environment is
- Visual Studio Code 1.36.0
- VIM 1.8.1
- Windows 10 Pro 1809

**Comment by muellerto** [comment]
Update: the cursor movement issue seems to be related to tabulators. There's already a case for that.

**Comment by pomdtr** [comment]
Remapping existing operator is also inconsistent:
```json
    "vim.visualModeKeyBindingsNonRecursive": [
        {
            "before": [
                ">"
            ],
            "commands": [
                "editor.action.indentLines"
            ]
        },
        {
            "before": [
                "<"
            ],
            "commands": [
                "editor.action.outdentLines"
            ]
        },
    ]
```
Doesn't work half of the time

**Comment by J-Fields** [comment]
>I also detected that cursor movement in normal mode is also not really good, especially up and down. Try this in your settings.json file: go onto the first letter of any line, this is the character right of a " and then move up and down using the arrow keys or j and k. The cursor should always stay in the same column but it doesn't. On some lines the cursor jumps three characters to the right. Why? These lines seem to be always the same lines, so they must have a hidden criterion which leads to the jump, but what is it?

Are you using tabs? If so, this is a known issue - we don't always take into account the configured tab size.

**Comment by muellerto** [comment]
Yes I use tabs. But I already said this problem seems to have other cases. Ignore this remark in my original post.

The **main problem** is that I can't remap cursor keys. This makes indeed VSCodeVim difficult to use. I always hack commands into my text. When I detect this I must hit Esc, remove the crap, repsosition the cursor and type all that again in normal mode. This costs a lot of time.

You could say it's all related to me, I could use the keys everyone uses. I want to answer it's related to cursor movement in insert mode. This is not vi behavior. Indeed it's rather a problem in an editor having modes. I found a clever solution for that problem which is practicable in almost every vi implementation on every OS even on very silly terminals, but VSCodeVim stumbles on that.

At the moment I use this Colmak-Vim extension which seems to be a fork of VSCodeVim, don't know. Perhaps this is older and has other disadvantages **but** it remaps cursor keys much better.
